web/src/copy_and_paste.ts

```diff
@@ -510,6 +531,154 @@ export function paste_handler_converter(paste_html: string): string {
+        Newlines in markdown are translated into <br> instead of empty katex-displays here.
+
+        For more information:
+        https://chat.zulip.org/#narrow/channel/9-issues/topic/Replying.20to.20highlighted.20text.2C.20LaTeX.20is.20not.20preserved.20.2331608/near/1991687
+    */
+
+    const processed_math_block_parents = new Set();
+    const processed_inline_math_block_parents = new Set();
+    turndownService.addRule("katex-math-block", {
+        filter(node) {
+            if (
+                node.classList.contains("katex-display") &&
+                !node.closest(".katex-display > .katex-display")
+            ) {
+                return true;
+            }
+
+            return false;
+        },
+        replacement(content: string, node) {
+            assert(node instanceof HTMLElement);
+            let math_block_markdown = "```math\n";
+            const parent = node.parentElement!;
+            if (processed_math_block_parents.has(parent)) {
+                return "";
+            }
+            processed_math_block_parents.add(parent);
+            let consecutive_empty_display_count = 0;
+            for (const child of parent.children) {
+                const annotation_element = child.querySelector(
+                    '.katex-mathml annotation[encoding="application/x-tex"]',
+                );
+                if (annotation_element?.textContent) {
+                    const katex_source = annotation_element.textContent.trim();
+                    math_block_markdown += katex_source + "\n\n";
+                    consecutive_empty_display_count = 0;
+                } else {
+                    // Handling cases where math block is selected directly without any preceding text.
+                    // The initial katex display doesn't have the annotation when selection is done in this manner.
+                    if (
+                        child.classList.contains("katex-display") &&
+                        child.querySelector("math")?.textContent !== ""
+                    ) {
+                        math_block_markdown += content;
+                        continue;
+                    }
+                    if (consecutive_empty_display_count === 0) {
+                        math_block_markdown += "\n\n\n";
+                    } else {
+                        math_block_markdown += "\n\n";
+                    }
+                    consecutive_empty_display_count += 1;
+                }
+            }
+            math_block_markdown = math_block_markdown.slice(0, -1); // Don't add extra newline add the end
+            return (math_block_markdown += "```");
+        },
+    });
+
+    turndownService.addRule("katex-inline-math", {
+        filter(node) {
+            // Allow the intermediate text blocks, so that they don't get appended to the end.
+            if (node.classList.contains("zulip-paste-parser-text-node")) {
+                return true;
+            }
```

Comment:
What about other inline nodes that may be inside a `p` tag, like a bold block? Does those end up appended at the end in some weird way?

Reply:
I think the theory is we don't need to do anything, because they'll be naturally in the child list?